File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/telemetry/KeystoreTelemetryEmitter.java
package org.hyperledger.iroha.android.telemetry;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import org.hyperledger.iroha.android.IrohaKeyManager;
import org.hyperledger.iroha.android.crypto.KeyGenerationOutcome;
import org.hyperledger.iroha.android.crypto.KeyProviderMetadata;
import org.hyperledger.iroha.android.crypto.keystore.attestation.AttestationResult;

/** Emits Android keystore attestation telemetry events. */
public final class KeystoreTelemetryEmitter {

  private static final KeystoreTelemetryEmitter NOOP =
      new KeystoreTelemetryEmitter(null, TelemetryOptions.Redaction.disabled(), DeviceProfileProvider.disabled());

  private final TelemetrySink sink;
  private final TelemetryOptions.Redaction redaction;
  private final DeviceProfileProvider deviceProfileProvider;

  private KeystoreTelemetryEmitter(
      final TelemetrySink sink,
      final TelemetryOptions.Redaction redaction,
      final DeviceProfileProvider deviceProfileProvider) {
    this.sink = sink;
    this.redaction = redaction;
    this.deviceProfileProvider = deviceProfileProvider;
  }

  /** Returns a no-op emitter. */
  public static KeystoreTelemetryEmitter noop() {
    return NOOP;
  }

  /**
   * Creates an emitter backed by {@link TelemetryOptions} and {@link TelemetrySink}. Returns a
   * no-op instance when telemetry is disabled or the sink is missing.
   */
  public static KeystoreTelemetryEmitter from(
      final TelemetryOptions options,
      final TelemetrySink sink,
      final DeviceProfileProvider provider) {
    if (options == null || sink == null || !options.enabled()) {
      return noop();
    }
    final DeviceProfileProvider safeProvider =
        provider == null ? DeviceProfileProvider.disabled() : provider;
    return new KeystoreTelemetryEmitter(sink, options.redaction(), safeProvider);
  }

  /** Records a successful attestation verification/generation. */
  public void recordResult(
      final String alias, final KeyProviderMetadata metadata, final AttestationResult result) {
    if (sink == null || redaction == null || result == null) {
      return;
    }
    final Optional<String> aliasLabel = aliasLabel(alias);
    if (aliasLabel.isEmpty()) {
      return;
    }
    final String securityLevel = result.attestationSecurityLevel().name().toLowerCase(Locale.ROOT);
    final String digest = leafCertificateDigest(result);
    final String brandBucket = deviceBrandBucket();
    sink.emitSignal(
        "android.keystore.attestation.result",
        Map.of(
            "alias_label", aliasLabel.get(),
            "security_level", securityLevel,
            "attestation_digest", digest,
            "device_brand_bucket", brandBucket,
            "provider", metadata == null ? "unknown" : metadata.name()));
  }

  /**
   * Records a key generation event, including the preference requested and the hardware route used.
   */
  public void recordKeyGeneration(
      final String alias,
      final IrohaKeyManager.KeySecurityPreference preference,
      final KeyProviderMetadata metadata,
      final KeyGenerationOutcome.Route route,
      final boolean fallback) {
    if (sink == null || redaction == null) {
      return;
    }
    final Optional<String> aliasLabel = aliasLabel(alias);
    if (aliasLabel.isEmpty()) {
      return;
    }
    final String preferenceLabel =
        preference == null ? "unknown" : preference.name().toLowerCase(Locale.ROOT);
    final String routeLabel = route == null ? "unknown" : route.name().toLowerCase(Locale.ROOT);
    sink.emitSignal(
        "android.keystore.keygen",
        Map.of(
            "alias_label", aliasLabel.get(),
            "preference", preferenceLabel,
            "route", routeLabel,
            "fallback", fallback,
            "provider", metadata == null ? "unknown" : metadata.name(),
            "device_brand_bucket", deviceBrandBucket()));
  }

  /**
   * Records a failure to validate Ed25519 SPKI outputs from a key provider.
   */
  public void recordKeyValidationFailure(
      final String alias,
      final IrohaKeyManager.KeySecurityPreference preference,
      final KeyProviderMetadata metadata,
      final String phase,
      final String reason,
      final int spkiLength,
      final int expectedLength,
      final String spkiPrefixHex) {
    if (sink == null || redaction == null) {
      return;
    }
    final Optional<String> aliasLabel = aliasLabel(alias);
    if (aliasLabel.isEmpty()) {
      return;
    }
    final String preferenceLabel =
        preference == null ? "unknown" : preference.name().toLowerCase(Locale.ROOT);
    final String phaseLabel = phase == null || phase.isBlank() ? "unknown" : phase;
    final String failureReason = reason == null || reason.isBlank() ? "unknown" : reason;
    final String prefix = spkiPrefixHex == null || spkiPrefixHex.isBlank() ? "unknown" : spkiPrefixHex;
    sink.emitSignal(
        "android.keystore.key_validation.failure",
        Map.of(
            "alias_label", aliasLabel.get(),
            "preference", preferenceLabel,
            "provider", metadata == null ? "unknown" : metadata.name(),
            "phase", phaseLabel,
            "reason", failureReason,
            "spki_length", spkiLength,
            "expected_spki_length", expectedLength,
            "spki_prefix", prefix,
            "device_brand_bucket", deviceBrandBucket()));
  }

  /** Records an attestation failure surfaced by {@link AttestationVerificationException}. */
  public void recordFailure(
      final String alias, final KeyProviderMetadata metadata, final String reason) {
    if (sink == null || redaction == null) {
      return;
    }
    final Optional<String> aliasLabel = aliasLabel(alias);
    if (aliasLabel.isEmpty()) {
      return;
    }
    final String failureReason = reason == null || reason.isBlank() ? "unknown" : reason;
    sink.emitSignal(
        "android.keystore.attestation.failure",
        Map.of(
            "alias_label", aliasLabel.get(),
            "failure_reason", failureReason,
            "provider", metadata == null ? "unknown" : metadata.name()));
  }


  private Optional<String> aliasLabel(final String alias) {
    if (alias == null || alias.isBlank()) {
      return Optional.empty();
    }
    if (!redaction.enabled()) {
      return Optional.empty();
    }
    return redaction.hashIdentifier(alias);
  }

  private String deviceBrandBucket() {
    return deviceProfileProvider
        .snapshot()
        .map(DeviceProfile::bucket)
        .orElse("unknown");
  }

  private static String leafCertificateDigest(final AttestationResult result) {
    try {
      final MessageDigest digest = MessageDigest.getInstance("SHA-256");
      final byte[] encoded = result.leafCertificate().getEncoded();
      final byte[] hash = digest.digest(encoded);
      return bytesToHex(hash);
    } catch (final NoSuchAlgorithmException | java.security.cert.CertificateEncodingException ex) {
      return "error";
    }
  }

  private static String bytesToHex(final byte[] bytes) {
    final StringBuilder builder = new StringBuilder(bytes.length * 2);
    for (final byte value : bytes) {
      builder.append(String.format(Locale.ROOT, "%02x", value));
    }
    return builder.toString();
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/telemetry/KeystoreTelemetryEmitter.kt
package org.hyperledger.iroha.sdk.telemetry

import java.security.cert.X509Certificate
import java.security.MessageDigest
import java.util.Locale
import org.hyperledger.iroha.sdk.crypto.KeyGenerationOutcome
import org.hyperledger.iroha.sdk.crypto.KeyProviderMetadata

/** Emits Android keystore attestation telemetry events. */
class KeystoreTelemetryEmitter private constructor(
    private val sink: TelemetrySink?,
    private val redaction: Redaction?,
    private val deviceProfileProvider: DeviceProfileProvider,
) {
    /**
     * Records a successful attestation verification/generation.
     *
     * @param securityLevel label for the attestation security level (e.g. "software", "tee", "strongbox")
     * @param leafCertificate the leaf certificate from the attestation chain
     */
    fun recordResult(alias: String?, metadata: KeyProviderMetadata?, securityLevel: String?, leafCertificate: X509Certificate?) {
        if (sink == null || redaction == null || securityLevel == null) return
        val aliasLabel = aliasLabel(alias) ?: return
        val digest = leafCertificateDigest(leafCertificate)
        val brandBucket = deviceBrandBucket()
        sink.emitSignal(
            "android.keystore.attestation.result",
            mapOf(
                "alias_label" to aliasLabel,
                "security_level" to securityLevel.lowercase(Locale.ROOT),
                "attestation_digest" to digest,
                "device_brand_bucket" to brandBucket,
                "provider" to (metadata?.name ?: "unknown"),
            ),
        )
    }

    /** Records a key generation event, including the preference requested and the hardware route used. */
    fun recordKeyGeneration(
        alias: String?,
        preferenceLabel: String?,
        metadata: KeyProviderMetadata?,
        route: KeyGenerationOutcome.Route?,
        fallback: Boolean,
    ) {
        if (sink == null || redaction == null) return
        val aliasLabel = aliasLabel(alias) ?: return
        val routeLabel = route?.name?.lowercase(Locale.ROOT) ?: "unknown"
        sink.emitSignal(
            "android.keystore.keygen",
            mapOf(
                "alias_label" to aliasLabel,
                "preference" to (preferenceLabel?.lowercase(Locale.ROOT) ?: "unknown"),
                "route" to routeLabel,
                "fallback" to fallback,
                "provider" to (metadata?.name ?: "unknown"),
                "device_brand_bucket" to deviceBrandBucket(),
            ),
        )
    }

    /** Records a failure to validate Ed25519 SPKI outputs from a key provider. */
    fun recordKeyValidationFailure(
        alias: String?,
        preferenceLabel: String?,
        metadata: KeyProviderMetadata?,
        phase: String?,
        reason: String?,
        spkiLength: Int,
        expectedLength: Int,
        spkiPrefixHex: String?,
    ) {
        if (sink == null || redaction == null) return
        val aliasLabel = aliasLabel(alias) ?: return
        val phaseLabel = if (phase.isNullOrBlank()) "unknown" else phase
        val failureReason = if (reason.isNullOrBlank()) "unknown" else reason
        val prefix = if (spkiPrefixHex.isNullOrBlank()) "unknown" else spkiPrefixHex
        sink.emitSignal(
            "android.keystore.key_validation.failure",
            mapOf(
                "alias_label" to aliasLabel,
                "preference" to (preferenceLabel?.lowercase(Locale.ROOT) ?: "unknown"),
                "provider" to (metadata?.name ?: "unknown"),
                "phase" to phaseLabel,
                "reason" to failureReason,
                "spki_length" to spkiLength,
                "expected_spki_length" to expectedLength,
                "spki_prefix" to prefix,
                "device_brand_bucket" to deviceBrandBucket(),
            ),
        )
    }

    /** Records an attestation failure. */
    fun recordFailure(alias: String?, metadata: KeyProviderMetadata?, reason: String?) {
        if (sink == null || redaction == null) return
        val aliasLabel = aliasLabel(alias) ?: return
        val failureReason = if (reason.isNullOrBlank()) "unknown" else reason
        sink.emitSignal(
            "android.keystore.attestation.failure",
            mapOf(
                "alias_label" to aliasLabel,
                "failure_reason" to failureReason,
                "provider" to (metadata?.name ?: "unknown"),
            ),
        )
    }

    private fun aliasLabel(alias: String?): String? {
        if (alias.isNullOrBlank()) return null
        if (redaction == null || !redaction.enabled) return null
        return redaction.hashIdentifier(alias).orElse(null)
    }

    private fun deviceBrandBucket(): String =
        deviceProfileProvider.snapshot()
            .map { it.bucket }
            .orElse("unknown")

    companion object {
        private val NOOP = KeystoreTelemetryEmitter(null, Redaction.disabled(), DeviceProfileProvider.disabled())

        @JvmStatic
        fun noop(): KeystoreTelemetryEmitter = NOOP

        @JvmStatic
        fun from(
            options: TelemetryOptions?,
            sink: TelemetrySink?,
            provider: DeviceProfileProvider?,
        ): KeystoreTelemetryEmitter {
            if (options == null || sink == null || !options.enabled) return noop()
            val safeProvider = provider ?: DeviceProfileProvider.disabled()
            return KeystoreTelemetryEmitter(sink, options.redaction, safeProvider)
        }

        private fun leafCertificateDigest(cert: X509Certificate?): String = try {
            if (cert == null) "unknown"
            else {
                val digest = MessageDigest.getInstance("SHA-256")
                bytesToHex(digest.digest(cert.encoded))
            }
        } catch (_: Exception) {
            "error"
        }

        private fun bytesToHex(bytes: ByteArray): String =
            bytes.joinToString("") { "%02x".format(it) }
    }
}
